[docs] Add CSI modules architecture section (#19003)
deckhouse/deckhouse: docs/documentation/pages/architecture/storage/CSI_DRIVER_SCSI_GENERIC.md

---
title: SCSI Generic CSI driver
permalink: en/architecture/storage/csi-drivers/csi-driver-scsi-generic.html
search: csi-scsi-generic, csi driver
description: Architecture of the csi-scsi-generic CSI driver in Deckhouse Kubernetes Platform.
---

The `csi-scsi-generic` CSI driver is the implementation of the [Container Storage Interface (CSI)](https://github.com/container-storage-interface/spec/blob/master/spec.md) standard to manage volumes on SCSI-connected storage systems in Deckhouse Kubernetes Platform (DKP).

## Driver architecture

{% alert level="info" %}
The following simplifications are made in the diagram:

* The diagram shows containers in different pods interacting directly with each other. In reality, they communicate via the corresponding Kubernetes Services (internal load balancers). Service names are omitted if they are obvious from the diagram context. Otherwise, the Service name is shown above the arrow.
* Pods may run multiple replicas. However, each pod is shown as a single replica in the diagram.
{% endalert %}

> The Level 2 C4 architecture of the `csi-scsi-generic` CSI driver and its interactions with other components of Deckhouse Kubernetes Platform (DKP) are shown in the following diagram:

Reviewer: ```suggestion
The Level 2 C4 architecture of the `csi-scsi-generic` CSI driver and its interactions with other components of DKP are shown in the following diagram:
```

   The plugin has privileged access to the filesystem of each node. On Linux, this requires the `CAP_SYS_ADMIN` capability. This is necessary to perform mount operations and interact with block devices.
   {% endalert %}

   It consists of the following containers:

   * **scsi-modules-loader**: Init container that loads the kernel module required for iSCSI (`iscsi_tcp`).

   * **node**: Main container implementing CSI driver functionality through the gRPC services Identity Service and Node Service according to the [CSI specification](https://github.com/container-storage-interface/spec/blob/master/spec.md#rpc-interface).

   * **node-driver-registrar**: Sidecar container that registers the Node Plugin with [kubelet](../../kubernetes-and-scheduling/kubelet.html). It calls the RPC methods `GetPluginInfo` and `NodeGetInfo` in the node container to retrieve plugin and node information. Communication with the node container occurs over gRPC via a Unix socket.

## Driver interactions

The driver interacts with the following components:

1. **Kube-apiserver**: Watches PersistentVolume, PersistentVolumeClaim, VolumeAttachment, PendingResizeRequest, SCSIDevice, and SCSITarget resources.

> 1. **SCSI-connected storage systems**: Orchestrates the use of already available SCSI devices, including their binding/cleanup and node attachment.

Reviewer: ```suggestion
1. **SCSI-connected storage systems**: Coordinates the usage of already available SCSI devices, including their binding/cleanup and node attachment.
```

The following external components interact with the driver:

1. [Kubelet](../../kubernetes-and-scheduling/kubelet.html):

* Checks CSI driver livenessProbe.
* Registers the Node Plugin.

> * Calls `NodeStageVolume`, `NodeUnstageVolume`, `NodePublishVolume`, and `NodeUnpublishVolume` RPCs in the Node Plugin.

Reviewer: ```suggestion
   * Checks CSI driver livenessProbe.
   * Registers the Node Plugin.
   * Calls `NodeStageVolume`, `NodeUnstageVolume`, `NodePublishVolume`, and `NodeUnpublishVolume` RPCs in the Node Plugin.
```
